sql/migrate/migrate.go

```diff
@@ -436,79 +418,111 @@ func (e *Executor) Pending(ctx context.Context) ([]File, error) {
 			return nil, fmt.Errorf("sql/migrate: execute: revisions and migrations mismatch: rev %q <> file %q", v, r.Version)
 		}
 	}
-	if len(migrations) == len(revs) {
+	// If there are no revisions yet, return all migration files as pending.
+	if len(revs) == 0 {
+		return migrations, nil
+	}
+	pending := migrations[len(revs):]
+	// If the last revision is not applied in full, it is considered pending as well.
+	if last := revs[len(revs)-1]; last.Applied < last.Total {
+		pending = append(migrations[len(revs)-1:len(revs)], pending...)
+	}
+	if len(pending) == 0 {
 		return nil, ErrNoPendingFiles
 	}
-	return migrations[len(revs):], nil
+	return pending, nil
 }
 
 // Execute executes the given migration file on the database. It does not check for the database to be clean before
 // attempting to apply the changes. This behavior is required to enabled "fixing" a broken state.
 func (e *Executor) Execute(ctx context.Context, m File) (err error) {
-	r := &Revision{ExecutedAt: time.Now(), ExecutionState: StateOngoing}
-	// Make sure to store the Revision information.
-	defer func(ctx context.Context, rrw RevisionReadWriter, r *Revision) {
-		if err2 := e.rrw.WriteRevision(ctx, r); err2 != nil {
-			err = wrap(fmt.Errorf("execute: write revision: %w", err2), err)
-		}
-	}(ctx, e.rrw, r)
-	sc := e.dir.(Scanner)
-	r.Version, err = sc.Version(m)
+	hf, err := HashSum(e.dir)
 	if err != nil {
-		return r.setGoErr(fmt.Errorf("sql/migrate: execute: scan version from %q: %w", m.Name(), err))
+		return fmt.Errorf("sql/migrate: execute: compute hash: %w", err)
 	}
-	r.Description, err = sc.Desc(m)
+	version, err := e.dir.(Scanner).Version(m)
 	if err != nil {
-		return r.setGoErr(fmt.Errorf("sql/migrate: execute: scan description from %q: %w", m.Name(), err))
-
-	}
-	if e.log != nil {
-		e.log.Log(LogFile{r.Version, r.Description})
+		return fmt.Errorf("sql/migrate: execute: scan version from %q: %w", m.Name(), err)
 	}
-	hf, err := HashSum(e.dir)
+	desc, err := e.dir.(Scanner).Desc(m)
 	if err != nil {
-		return fmt.Errorf("sql/migrate: execute: create hash file: %w", err)
+		return fmt.Errorf("sql/migrate: execute: scan description from %q: %w", m.Name(), err)
 	}
-	r.Hash, err = hf.sumByName(m.Name())
+	hash, err := hf.sumByName(m.Name())
 	if err != nil {
-		return r.setGoErr(fmt.Errorf("sql/migrate: execute: scanning checksum for file %q: %w", m.Name(), err))
+		return fmt.Errorf("sql/migrate: execute: scanning checksum from %q: %w", m.Name(), err)
 	}
-	stmts, err := sc.Stmts(m)
+	stmts, err := e.dir.(Scanner).Stmts(m)
 	if err != nil {
-		return r.setGoErr(fmt.Errorf("sql/migrate: execute: scanning statements from file %q: %w", m.Name(), err))
+		return fmt.Errorf("sql/migrate: execute: scanning statements from %q: %w", m.Name(), err)
+	}
+	// If there already is a revision with this version in the database,
+	// and it is partially applied, continue where the last attempt was left off.
+	r, err := e.rrw.ReadRevision(ctx, version)
```

Comment:
Don't we want to ensure the file wasn't changed? e.g. compare hashes?

In case it was changed, should we fail or apply it from the first statement?